### ISSUE TITLE
Building and linking Android library

### ISSUE THREAD

**Issue by spring-reverb** [clostra/newnode#111]
We're trying to build newnode from source code for Android, but the application with linked library does not start.

Can you please provide the official guide for building and linking Android version of library from publicly available source code?

Our process at the moment:

**Building**

```shell
% arch -x86_64 ./build.android.sh
-rwxr-xr-x  1 sr  staff   2208036 Jan 17 15:53 android/src/main/jniLibs/armeabi-v7a/libnewnode.so
-rw-r--r--  1 sr  staff  12018417 Jan 17 15:53 android/src/main/jniLibs/armeabi-v7a/mapping.txt
-rwxr-xr-x  1 sr  staff   2278856 Jan 17 15:53 android/src/main/jniLibs/arm64-v8a/libnewnode.so
-rw-r--r--  1 sr  staff  10709944 Jan 17 15:53 android/src/main/jniLibs/arm64-v8a/mapping.txt
-rwxr-xr-x  1 sr  staff   2475712 Jan 17 15:53 android/src/main/jniLibs/x86_64/libnewnode.so
-rw-r--r--  1 sr  staff  14478759 Jan 17 15:53 android/src/main/jniLibs/x86_64/mapping.txt

% ./gradlew build   

BUILD SUCCESSFUL in 831ms
61 actionable tasks: 61 up-to-date
```

**Linking**

```kotlin
implementation files('../../../newnode/android/build/outputs/aar/newnode-release.aar')
```

**Comment by spring-reverb** [comment]
BTW, "successful" build is the result of our changes to codebase (see below).


Here what happens with sources from master branch:

```
% arch -x86_64 ./build.android.sh && cd android && arch -x86_64 ./gradlew build
<…>
android.c:313:19: fatal error: suggest braces around initialization of subobject [-Wmissing-braces]
    endpoint e = {0};
                  ^
                  {}
1 error generated.
```

We can fix it by replacing in `android.c` this code:

```
endpoint e = {0};
```

with:

```
endpoint e = {};
```

After that we receive a new error:

```
<…>
tarting a Gradle Daemon, 1 incompatible Daemon could not be reused, use --status for details

> Task :lintDebug FAILED
Wrote HTML report to file:///Users/sr/Coding/reverb/newnode_vanilla_take_two/android/build/reports/lint-results-debug.html
Lint found 5 errors and 4 warnings. First failure:
/Users/sr/Coding/reverb/newnode_vanilla_take_two/android/src/main/AndroidManifest.xml:16: Error: Class referenced in the manifest, com.clostra.newnode.internal.NearbyHelper.BatteryLevelReceiver, was not found in the project or the libraries [MissingClass]
        <receiver android:name=".internal.NearbyHelper.BatteryLevelReceiver" android:exported="true">
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Lint found errors in the project; aborting build.

<…>
```

This error can be fixed by replacing this string if `AndroidManifest.xml`:

```
<receiver android:name=".internal.NearbyHelper.BatteryLevelReceiver" android:exported="true">
```

with

```
<receiver android:name=".internal.NewNode$BatteryLevelReceiver" android:exported="true">
```

After that we have yet another error:

```
<…>
> Task :lintDebug FAILED
Wrote HTML report to file:///Users/sr/Coding/reverb/newnode_vanilla_take_two/android/build/reports/lint-results-debug.html
Lint found 4 errors and 4 warnings. First failure:
/Users/sr/Coding/reverb/newnode_vanilla_take_two/android/src/main/java/com/clostra/newnode/internal/NewNode.java:259: Error: Call requires API level 29 (current min is 16): sendPacket [NewApi]
            bluetooth.sendPacket(packet, endpoint);
                      ~~~~~~~~~~

Lint found errors in the project; aborting build.
<…>
```

And we can fix that by adding suppress to linter for particular methods in `internal/NewNode.java`, cause it's a safe thing for this particular case:


`import android.annotation.SuppressLint;
`
```
@SuppressLint("NewApi")
    void sendPacket(byte[] packet, byte[] endpoint) {
```

```
@SuppressLint("NewApi")
    static void startNearby() {
```

```
@SuppressLint("NewApi")
    static void stopNearby() {
```

And only after that we get the "success" in building process, which does not work with an actual app:

```
Launching lib/main.dart on Redmi 4 in debug mode...
Running Gradle task 'assembleDebug'...
✓  Built build/app/outputs/flutter-apk/app-debug.apk.
Installing build/app/outputs/flutter-apk/app.apk...
E/AndroidRuntime(24049): FATAL EXCEPTION: main
E/AndroidRuntime(24049): Process: io.reverb.reverb, PID: 24049
E/AndroidRuntime(24049): java.lang.NoClassDefFoundError: Failed resolution of: Lcom/bugsnag/android/Configuration;
E/AndroidRuntime(24049): 	at com.clostra.newnode.internal.NewNode.bugsnagClientInit(NewNode.java:266)
E/AndroidRuntime(24049): 	at com.clostra.newnode.internal.NewNode.init(NewNode.java:209)
E/AndroidRuntime(24049): 	at com.clostra.newnode.NewNode.init(NewNode.java:88)
E/AndroidRuntime(24049): 	at io.reverb.reverb.MainActivity.onCreate(MainActivity.java:27)
E/AndroidRuntime(24049): 	at android.app.Activity.performCreate(Activity.java:6355)
E/AndroidRuntime(24049): 	at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1108)
E/AndroidRuntime(24049): 	at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2438)
E/AndroidRuntime(24049): 	at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2545)
E/AndroidRuntime(24049): 	at android.app.ActivityThread.access$1100(ActivityThread.java:151)
E/AndroidRuntime(24049): 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1396)
E/AndroidRuntime(24049): 	at android.os.Handler.dispatchMessage(Handler.java:102)
E/AndroidRuntime(24049): 	at android.os.Looper.loop(Looper.java:157)
E/AndroidRuntime(24049): 	at android.app.ActivityThread.main(ActivityThread.java:5601)
E/AndroidRuntime(24049): 	at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime(24049): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:774)
E/AndroidRuntime(24049): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:652)
E/AndroidRuntime(24049): Caused by: java.lang.ClassNotFoundException: Didn't find class "com.bugsnag.android.Configuration" on path: DexPathList[[zip file "/data/app/io.reverb.reverb-1/base.apk"],nativeLibraryDirectories=[/data/app/io.reverb.reverb-1/lib/arm64, /data/app/io.reverb.reverb-1/base.apk!/lib/arm64-v8a, /vendor/lib64, /system/lib64]]
E/AndroidRuntime(24049): 	at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
E/AndroidRuntime(24049): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
E/AndroidRuntime(24049): 	at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
E/AndroidRuntime(24049): 	... 16 more
E/AndroidRuntime(24049): 	Suppressed: java.lang.ClassNotFoundException: com.bugsnag.android.Configuration
E/AndroidRuntime(24049): 		at java.lang.Class.classForName(Native Method)
E/AndroidRuntime(24049): 		at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
E/AndroidRuntime(24049): 		at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
E/AndroidRuntime(24049): 		at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
E/AndroidRuntime(24049): 		... 17 more
E/AndroidRuntime(24049): 	Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available

```

We can comment out the bugsnag stuff:

```
// bugsnagClientInit();
```

and

```
void bugsnagClientInit() {
//        Configuration config = new Configuration("141ea25aa72c276c49d3a154b82f2b1f");
//        config.setAppVersion(VERSION);
//        config.setBuildUUID(VERSION);
//        config.setSendThreads(true);
//        config.setPersistUserBetweenSessions(true);
//        config.setAutoCaptureSessions(true);
//        config.setEnableExceptionHandler(true);
//
//        bugsnagClient = new Client(app(), config);
//        bugsnagClient.setProjectPackages("com.clostra.newnode");
//        //bugsnagClient.setLoggingEnabled(true);
//        NativeInterface.setClient(bugsnagClient);
//
//        bugsnagClient.addObserver(new BugsnagObserver());
//        updateBugsnagDetails(NotifyType.ALL.getValue());
    }
```

But the application fails to start with final error:

```
<…>

2022-02-09 17:03:41.784 26085-26085/io.reverb.reverb D/newnode: Loading built-in NewNodeInternal
2022-02-09 17:03:41.786 26085-26085/io.reverb.reverb D/NearbyHelper: loading built-in newnode
2022-02-09 17:03:41.788 26085-26148/io.reverb.reverb I/flutter: Observatory listening on http://127.0.0.1:43011/eh3XoS0NuAk=/
2022-02-09 17:03:41.793 26085-26085/io.reverb.reverb D/NearbyHelper: batteryPct: 100.0
2022-02-09 17:03:41.794 26085-26085/io.reverb.reverb D/newnode: application id io.reverb.reverb
2022-02-09 17:03:41.794 26085-26085/io.reverb.reverb V/newnode: getrlimit: r:0 cur:4096 max:4096
2022-02-09 17:03:42.529 26085-26168/io.reverb.reverb V/newnode: [0209/170342.510067:ERROR:elf_dynamic_array_reader.h(64)] tag not found
    [0209/170342.525507:ERROR:elf_dynamic_array_reader.h(64)] tag not found
2022-02-09 17:03:42.755 26085-26085/io.reverb.reverb A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x30 in tid 26085 (o.reverb.reverb)
```

This error tracks down to last call to `getrlimit` which is called in `set_max_nofile`. It seems that the method does not finish, cause it should print `getrlimit` twice:

```
void set_max_nofile()
{
    rlimit nofile;
    int r = getrlimit(RLIMIT_NOFILE, &nofile);
    debug("getrlimit: r:%d cur:%zu max:%zu\n", r, (size_t)nofile.rlim_cur, (size_t)nofile.rlim_max);
    for (rlim_t max = nofile.rlim_max; ;) {
        rlim_t mid = (max - nofile.rlim_cur) / 2;
        if (!mid) {
            break;
        }
        nofile.rlim_cur += mid;
        r = setrlimit(RLIMIT_NOFILE, &nofile);
        if (r) {
            max = nofile.rlim_cur;
            nofile.rlim_cur -= mid;
        }
    }
    r = getrlimit(RLIMIT_NOFILE, &nofile);
    debug("getrlimit: r:%d cur:%zu max:%zu\n", r, (size_t)nofile.rlim_cur, (size_t)nofile.rlim_max);
}
```

We tried to comment out the whole code, but we are still getting new errors on the way:

```
2022-02-09 17:55:49.740 27866-27866/io.reverb.reverb D/newnode: Loading built-in NewNodeInternal
2022-02-09 17:55:49.742 27866-27866/io.reverb.reverb D/NearbyHelper: loading built-in newnode
2022-02-09 17:55:49.749 27866-27866/io.reverb.reverb D/NearbyHelper: batteryPct: 100.0
2022-02-09 17:55:49.750 27866-27866/io.reverb.reverb D/newnode: application id io.reverb.reverb
2022-02-09 17:55:49.751 27866-27948/io.reverb.reverb V/newnode: libevent method: epoll
2022-02-09 17:55:49.753 27866-27948/io.reverb.reverb V/newnode: listening on UDP: [::]:53328
2022-02-09 17:55:49.753 27866-27948/io.reverb.reverb V/newnode: lsd IP_ADD_MEMBERSHIP 19 No such device
2022-02-09 17:55:49.754 27866-27948/io.reverb.reverb V/newnode: listening on TCP: 127.0.0.1:42030,48489
2022-02-09 17:55:49.755 27866-27866/io.reverb.reverb E/NearbyHelper: version 2.0.2 started
2022-02-09 17:55:49.756 27866-27866/io.reverb.reverb I/art: Rejecting re-init on previously-failed class java.lang.Class<com.clostra.newnode.internal.NearbyHelper$1>
2022-02-09 17:55:49.756 27866-27866/io.reverb.reverb I/art: Rejecting re-init on previously-failed class java.lang.Class<com.clostra.newnode.internal.NearbyHelper$1>
2022-02-09 17:55:49.758 27866-27949/io.reverb.reverb V/newnode: connect_more_injectors injector_pref:1
2022-02-09 17:55:49.770 27866-27866/io.reverb.reverb E/NearbyHelper: NearbyHelper
    java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/android/gms/nearby/connection/Strategy;
        at com.clostra.newnode.internal.NearbyHelper.<init>(NearbyHelper.java:39)
        at com.clostra.newnode.internal.NewNode.init(NewNode.java:228)
        at com.clostra.newnode.NewNode.init(NewNode.java:88)
        at io.reverb.reverb.MainActivity.onCreate(MainActivity.java:27)
        at android.app.Activity.performCreate(Activity.java:6355)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1108)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2438)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2545)
        at android.app.ActivityThread.access$1100(ActivityThread.java:151)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1396)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:157)
        at android.app.ActivityThread.main(ActivityThread.java:5601)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:774)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:652)
     Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.nearby.connection.Strategy" on path: DexPathList[[zip file "/data/app/io.reverb.reverb-2/base.apk"],nativeLibraryDirectories=[/data/app/io.reverb.reverb-2/lib/arm64, /data/app/io.reverb.reverb-2/base.apk!/lib/arm64-v8a, /vendor/lib64, /system/lib64]]
        at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
        at com.clostra.newnode.internal.NearbyHelper.<init>(NearbyHelper.java:39) 
        at com.clostra.newnode.internal.NewNode.init(NewNode.java:228) 
        at com.clostra.newnode.NewNode.init(NewNode.java:88) 
        at io.reverb.reverb.MainActivity.onCreate(MainActivity.java:27) 
        at android.app.Activity.performCreate(Activity.java:6355) 
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1108) 
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2438) 
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2545) 
        at android.app.ActivityThread.access$1100(ActivityThread.java:151) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1396) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:157) 
        at android.app.ActivityThread.main(ActivityThread.java:5601) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:774) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:652) 
    	Suppressed: java.lang.ClassNotFoundException: com.google.android.gms.nearby.connection.Strategy
        at java.lang.Class.classForName(Native Method)
        at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
        at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
        		... 17 more
     Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available
2022-02-09 17:55:49.775 27866-27929/io.reverb.reverb I/flutter: Observatory listening on http://127.0.0.1:53357/MONFFXBG6UY=/
2022-02-09 17:55:49.780 27866-27902/io.reverb.reverb I/FA: Tag Manager is not found and thus will not be used
2022-02-09 17:55:49.783 27866-27952/io.reverb.reverb D/OpenGLRenderer: Use EGL_SWAP_BEHAVIOR_PRESERVED: true
2022-02-09 17:55:49.818 27866-27866/io.reverb.reverb I/System.out: flutterEngine warmed up
2022-02-09 17:55:49.837 27866-27952/io.reverb.reverb I/OpenGLRenderer: Initialized EGL, version 1.4
2022-02-09 17:55:50.499 27866-27948/io.reverb.reverb V/newnode: [0209/175550.479866:ERROR:elf_dynamic_array_reader.h(64)] tag not found
    [0209/175550.495661:ERROR:elf_dynamic_array_reader.h(64)] tag not found
2022-02-09 17:55:50.667 27866-27866/io.reverb.reverb I/Choreographer: Skipped 50 frames!  The application may be doing too much work on its main thread.
2022-02-09 17:55:50.727 27866-27949/io.reverb.reverb A/libc: Fatal signal 11 (SIGSEGV), code 1, fault addr 0x30 in tid 27949 (newnode)
```

That looks like a rabbit hole and it would be great if you can provide some guidance/documentation on building library from source code.

**Comment by ghazel** [comment]
For multiple errors like this it would helpful to file multiple GitHub issues. Responses inline:

> ```
> % arch -x86_64 ./build.android.sh && cd android && arch -x86_64 ./gradlew build
> <…>
> android.c:313:19: fatal error: suggest braces around initialization of subobject [-Wmissing-braces]
>     endpoint e = {0};
> ```

This error does not occur for me. What is the NDK version and clang version, and the output of `uname -a`? Your compiler seems to be acting differently than the one I'm using.

> ```
> /Users/sr/Coding/reverb/newnode_vanilla_take_two/android/src/main/AndroidManifest.xml:16: Error: Class referenced in the manifest, com.clostra.newnode.internal.NearbyHelper.BatteryLevelReceiver, was not found in the project or the libraries [MissingClass]
> ```

This error is accurate, but I don't receive it when building. Did you set the linter settings in your env somehow? Similarly, your linter seems to be acting differently than the one I'm using.

> /Users/sr/Coding/reverb/newnode_vanilla_take_two/android/src/main/java/com/clostra/newnode/internal/NewNode.java:259: Error: Call requires API level 29 (current min is 16): sendPacket [NewApi]

Again, these linter errors do not occur for me.

> E/AndroidRuntime(24049): java.lang.NoClassDefFoundError: Failed resolution of: Lcom/bugsnag/android/Configuration;

This indicates you are not including the dependencies as in https://github.com/clostra/newnode/blob/b1e99d9aeeb910ea6ad5771d527a1d603c66e095/android/examples/WebViewSample/app/build.gradle#L26 which is recommended in the README here https://github.com/clostra/newnode/blob/b1e99d9aeeb910ea6ad5771d527a1d603c66e095/README.md?plain=1#L48

> But the application fails to start with final error:
> 
> This error tracks down to last call to `getrlimit` which is called in `set_max_nofile`. It seems that the method does not finish, cause it should print `getrlimit` twice:

What is the device type and android version?

> We tried to comment out the whole code, but we are still getting new errors on the way:
> 
> ```
> 2022-02-09 17:55:49.770 27866-27866/io.reverb.reverb E/NearbyHelper: NearbyHelper
>     java.lang.NoClassDefFoundError: Failed resolution of: Lcom/google/android/gms/nearby/connection/Strategy;
>         at com.clostra.newnode.internal.NearbyHelper.<init>(NearbyHelper.java:39)
>         at com.clostra.newnode.internal.NewNode.init(NewNode.java:228)
>         at com.clostra.newnode.NewNode.init(NewNode.java:88)
>         at io.reverb.reverb.MainActivity.onCreate(MainActivity.java:27)
>         at android.app.Activity.performCreate(Activity.java:6355)
>         at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1108)
>         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2438)
>         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2545)
>         at android.app.ActivityThread.access$1100(ActivityThread.java:151)
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1396)
>         at android.os.Handler.dispatchMessage(Handler.java:102)
>         at android.os.Looper.loop(Looper.java:157)
>         at android.app.ActivityThread.main(ActivityThread.java:5601)
>         at java.lang.reflect.Method.invoke(Native Method)
>         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:774)
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:652)
>      Caused by: java.lang.ClassNotFoundException: Didn't find class "com.google.android.gms.nearby.connection.Strategy" on path: DexPathList[[zip file "/data/app/io.reverb.reverb-2/base.apk"],nativeLibraryDirectories=[/data/app/io.reverb.reverb-2/lib/arm64, /data/app/io.reverb.reverb-2/base.apk!/lib/arm64-v8a, /vendor/lib64, /system/lib64]]
>         at dalvik.system.BaseDexClassLoader.findClass(BaseDexClassLoader.java:56)
>         at java.lang.ClassLoader.loadClass(ClassLoader.java:511)
>         at java.lang.ClassLoader.loadClass(ClassLoader.java:469)
>         at com.clostra.newnode.internal.NearbyHelper.<init>(NearbyHelper.java:39) 
>         at com.clostra.newnode.internal.NewNode.init(NewNode.java:228) 
>         at com.clostra.newnode.NewNode.init(NewNode.java:88) 
>         at io.reverb.reverb.MainActivity.onCreate(MainActivity.java:27) 
>         at android.app.Activity.performCreate(Activity.java:6355) 
>         at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1108) 
>         at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2438) 
>         at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2545) 
>         at android.app.ActivityThread.access$1100(ActivityThread.java:151) 
>         at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1396) 
>         at android.os.Handler.dispatchMessage(Handler.java:102) 
>         at android.os.Looper.loop(Looper.java:157) 
>         at android.app.ActivityThread.main(ActivityThread.java:5601) 
>         at java.lang.reflect.Method.invoke(Native Method) 
>         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:774) 
>         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:652) 
>     	Suppressed: java.lang.ClassNotFoundException: com.google.android.gms.nearby.connection.Strategy
>         at java.lang.Class.classForName(Native Method)
>         at java.lang.BootClassLoader.findClass(ClassLoader.java:781)
>         at java.lang.BootClassLoader.loadClass(ClassLoader.java:841)
>         at java.lang.ClassLoader.loadClass(ClassLoader.java:504)
>         		... 17 more
>      Caused by: java.lang.NoClassDefFoundError: Class not found using the boot class loader; no stack trace available
> ```

This is not an error, just a log message. However it is also due to not including all the dependencies.

**Comment by spring-reverb** [comment]
> What is the NDK version and clang version, and the output of uname -a?

NDK — 21.4.7075529, but we also tried these versions:

`21.4.7075529	22.1.7171670	23.1.7779620`

Clang:

`% arch -x86_64 clang --version
Apple clang version 13.0.0 (clang-1300.0.29.30)
Target: x86_64-apple-darwin21.2.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin`

(we are using `arch`, cause the processor is m1)

**Comment by spring-reverb** [comment]
`This error is accurate, but I don't receive it when building. Did you set the linter settings in your env somehow? Similarly, your linter seems to be acting differently than the one I'm using.`

I haven't setup any linter rules by myself. I think those are the defaults. It would be great to have suitable linter config file.

**Comment by spring-reverb** [comment]
> This indicates you are not including the dependencies as in
It makes sense. We'll try to include those dependencies in our gradle file. 

Thanks!

**Comment by spring-reverb** [comment]
> What is the device type and android version?

We tried on Xiaomi Redmi 4 (Android 6) and Android Emulator (Pixel 4 preset) with Android 11

**Comment by spring-reverb** [comment]
> This is not an error, just a log message. However it is also due to not including all the dependencies.

Well, the app does not start with that _log message_ 😄 . (But we'll try with dependencies)

**Comment by ghazel** [comment]
> > What is the NDK version and clang version, and the output of uname -a?
> 
> NDK — 21.4.7075529, but we also tried these versions:
> 
> `21.4.7075529 22.1.7171670 23.1.7779620`
> 
> Clang:
> 
> `% arch -x86_64 clang --version Apple clang version 13.0.0 (clang-1300.0.29.30) Target: x86_64-apple-darwin21.2.0 Thread model: posix InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin`
> 
> (we are using `arch`, cause the processor is m1)

`arch` could be related to the issue. I also compile on an M1, in a Rosetta terminal. Could you try that? (Be sure to clean build products first)

**Comment by spring-reverb** [comment]
>arch could be related to the issue. I also compile on an M1, in a Rosetta terminal. Could you try that? (Be sure to clean build products first)

So, I used rosetta terminal and added all the dependencies in out build.gradle.

After that we're having these issues:

```
2022-02-14 14:15:23.299 5775-5775/io.reverb.reverb D/newnode: application id io.reverb.reverb
2022-02-14 14:15:23.301 5775-5887/io.reverb.reverb V/newnode: libevent method: epoll
2022-02-14 14:15:23.302 5775-5775/io.reverb.reverb E/libc: Access denied finding property "net.dns1"
2022-02-14 14:15:23.302 5775-5775/io.reverb.reverb E/libc: Access denied finding property "net.dns2"
```

and fatal one:

```
2022-02-14 14:15:23.370 5775-5775/io.reverb.reverb D/Bluetooth: startScan
2022-02-14 14:15:23.370 5775-5775/io.reverb.reverb D/BluetoothAdapter: isLeEnabled(): ON
2022-02-14 14:15:23.374 5775-5798/io.reverb.reverb D/BluetoothLeScanner: onScannerRegistered() - status=0 scannerId=9 mScannerId=0
2022-02-14 14:15:23.374 5775-5890/io.reverb.reverb D/BluetoothGattServer: onServiceAdded() - handle=40 uuid=08076b03-983b-4154-93a5-4a6376b87993 status=0
2022-02-14 14:15:23.393 5775-5888/io.reverb.reverb A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x30 in tid 5888 (newnode), pid 5775 (o.reverb.reverb)

```

App still doesn't start (it starts with official build).
Full log in attachment: [logcat_rosetta_deps.log](https://github.com/clostra/newnode/files/8060613/logcat_rosetta_deps.log)

**Comment by ghazel** [comment]
> ```
> 2022-02-14 14:15:23.302 5775-5775/io.reverb.reverb E/libc: Access denied finding property "net.dns1"
> 2022-02-14 14:15:23.302 5775-5775/io.reverb.reverb E/libc: Access denied finding property "net.dns2"
> ```

This is https://github.com/clostra/newnode/issues/80 . Not fatal.

> and fatal one:
> 
> ```
> 2022-02-14 14:15:23.393 5775-5888/io.reverb.reverb A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x30 in tid 5888 (newnode), pid 5775 (o.reverb.reverb)
> ```

Narrowing this down and getting a stacktrace would be helpful. Again, make sure you're cleaning build products from the `arch` setup you had. Maybe start with a fresh checkout.

**Comment by spring-reverb** [comment]
Ok, I'll try with fresh checkout, but it will take some time

**Comment by spring-reverb** [comment]
TL;DR: We have the working build now!

So, here is the process again with fresh checkout.

First build (`./build.android.sh`) still fails with:

```
rm: *.o: No such file or directory
android.c:313:19: fatal error: suggest braces around initialization of subobject [-Wmissing-braces]
    endpoint e = {0};
                  ^
                  {}
1 error generated.
```

After fixes (`{0}` -> `{}`) `./build.android.sh` finishes successfully, but `./gradlew build` fails with:

```
% ./gradlew build

> Task :lintDebug FAILED
Wrote HTML report to file:///Users/sr/Coding/reverb/newnode_vanilla_take_three/android/build/reports/lint-results-debug.html
Lint found 5 errors and 4 warnings. First failure:
/Users/sr/Coding/reverb/newnode_vanilla_take_three/android/src/main/AndroidManifest.xml:16: Error: Class referenced in the manifest, com.clostra.newnode.internal.NearbyHelper.BatteryLevelReceiver, was not found in the project or the libraries [MissingClass]
        <receiver android:name=".internal.NearbyHelper.BatteryLevelReceiver" android:exported="true">
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Lint found errors in the project; aborting build.

Fix the issues identified by lint, or add the following to your build script to proceed with errors:
...
android {
    lintOptions {
        abortOnError false
    }
}
...

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':lintDebug'.
> A failure occurred while executing com.android.build.gradle.internal.lint.AndroidLintTask$AndroidLintLauncherWorkAction
   > There was a failure while executing work items
      > A failure occurred while executing com.android.build.gradle.internal.lint.AndroidLintWorkAction
         > Lint found errors in the project; aborting build.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 10s
61 actionable tasks: 61 executed
```

After fixes (`.internal.NewNode$BatteryLevelReceiver`) and lint-related changes (`@SuppressLint("NewApi")`) build successfully finishes:

```
> Task :lintDebug
Wrote HTML report to file:///Users/sr_work/Coding/reverb/newnode_vanilla_take_three/android/build/reports/lint-results-debug.html

BUILD SUCCESSFUL in 4s
61 actionable tasks: 19 executed, 42 up-to-date
```

Now we are trying to link the library to application by adding this to `build.gradle`:

```
    implementation files('../../../newnode_vanilla_take_three/android/build/outputs/aar/newnode-release.aar')
    // depdendencies for AAR-newnode:
    implementation 'com.bugsnag:bugsnag-android:4.7.0'
    implementation 'androidx.localbroadcastmanager:localbroadcastmanager:1.0.0'
    implementation 'androidx.appcompat:appcompat:1.4.1'
    implementation 'com.google.android.gms:play-services-nearby:18.+'
```

**And the app starts after that without any libc-related complaints!**

It would be great if you include some guides in README on how to build the library and share lint settings too, so that other developers can build the library. 

Thanks for your help and support!

**Comment by ghazel** [comment]
Your proposed fixes are accurate. We'll incorporate them. I don't see why our linter settings are different from yours. Everything should be in the repo.
